[#5808] fix(CLI): Fix improper exception throwing When a malformed name is passed to the CLI command

Abyss-lord opened the pull request:
### What changes were proposed in this pull request?

No exception should be thrown when a malformed name is passed to the CLI. Currently, passing a malformed name causes an IllegalNamespaceException. I’ve added error messages to inform the user when necessary arguments are missing.

Additionally, the `FullName.getNamePart()`  method no longer prints error messages, as the information it provides is limited. I think performing fine-grained argument validation in each method and providing specific hints is a better way to hint users.

### Why are the changes needed?

Fix: #5808 

### Does this PR introduce _any_ user-facing change?

NO

### How was this patch tested?

```bash
bin/gcli.sh table  list -i
# output: Missing required argument(s): METALAKE, CATALOG, SCHEMA

bin/gcli.sh table  list -i --metalake demo_metalake
# output: Missing required argument(s): CATALOG, SCHEMA

bin/gcli.sh table  list -i --metalake demo_metalake --name Hive_catalog
# output: Missing required argument(s): SCHEMA

bin/gcli.sh table  list -i --metalake demo_metalake --name Hive_catalog.default
# output: correct result
```

Abyss-lord commented on clients/cli/src/main/java/org/apache/gravitino/cli/FullName.java:
```diff
@@ -153,7 +153,7 @@ public String getNamePart(int position) {
     }
 
-    System.err.println(ErrorMessages.MISSING_NAME);
+    //    System.err.println(ErrorMessages.MISSING_NAME);
```
I'm unsure whether I should remove this code.

justinmclean replied:
It needs to be left in as that occurs when the name option missing

Abyss-lord commented:
@justinmclean Hi Justin, Could you please take a look when you have time?

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListTables.java:
```diff
@@ -48,6 +49,10 @@ public ListTables(
+    if (metalake == null || catalog == null || schema == null) {
+      showUnSpecifiedMessage();
+      return;
+    }
```
It might be better to add a method to the command base class and call that, as it is likely several commands need this? Just something to consider.

Abyss-lord replied:
> It might be better to add a method to the command base class and call that, as it is likely several commands need this? Just something to consider.

Thank you Justin. Can I modify the base classes freely? e.g. `Command` If so, I could move the argument handling logic into the Command class and update all the subclasses accordingly.

Abyss-lord replied:
> It might be better to add a method to the command base class and call that, as it is likely several commands need this? Just something to consider.

hi Justin, I noticed that the metalake parameter is almost never empty, it is given the information during the parsing phase, so do I need to check this parameter in the `handle` method?

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListTables.java:
```diff
@@ -64,6 +69,16 @@ public void handle() {
+      System.err.println(ErrorMessages.UNSPECIFIED_CATALOG);
+    } else {
+      System.err.println(ErrorMessages.UNSPECIFIED_SCHEMA);
+    }
```
Again a good candidate to move to base Command class I think.

Abyss-lord commented:
@justinmclean Hi Justin, Could you please take a look when you have time? I added two abstract methods to the Commands class: `hasCompleteArguments` and `displayMissingArguments` ,  and implemented them in all subclasses.
- `hasCompleteArguments`：The  method is used to check whether the current command has all the required arguments.
- `displayMissingArguments`：The  method is responsible for displaying information about the missing parameters.

justinmclean commented:
Sorry, but I would not do it this way, as it adds too much unneeded and duplicated code.

Taking one example that contains:
return CommandUtil.allNotNull(metalake, group);

a) there is not need to check metalake as that is already checked
b) there is no need to check group as it cannot be null

justinmclean commented:
There is also no need for code like this:

```
 @Override
   protected void displayMissingArguments() {
     // do nothing
   }
```

justinmclean commented:
One possible solution would be to add a method to check the name parameter and how many parts it should have. It may be that the best place to check for this is at the handleXXXCommand level?

Another is to add a single method (possibly overloaded) to the base Command that does the check and outputs error messages and only calls from the few commands that need it.

Abyss-lord commented:
> Sorry, but I would not do it this way, as it adds too much unneeded and duplicated code.
> 
> Taking one example that contains: return CommandUtil.allNotNull(metalake, group);
> 
> a) there is not need to check metalake as that is already checked b) there is no need to check group as it cannot be null

@justinmclean Maybe The create command should check metalake, in that way, we can fix #5825
My approach is to add a parameter check and output(maybe in one method) method at the Command level, and prevent the `handle` method from executing if the parameter validation fails.In this case,  a logical check will added to the `handle` method to ensure proper execution flow.

justinmclean commented:
That sounds like a better approach, and yes, checking metalake could give a better error message.

xunliu commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/CommandConstants.java:
I think we didn't define `CommandConstants.java`, we can directly use MetadataObject.Type in the `CommandUtil.java`, for example:
```
  public static String[] TABLE_MODEL =
      new String[] {
        MetadataObject.Type.METALKE.name(),
        MetadataObject.Type.CATALOG.name(),
        MetadataObject.Type.SCHEMA.name(),
        MetadataObject.Type.TABLE.name(),
        MetadataObject.Type.COLUMN.name()
      };
```

Abyss-lord replied:
thank you

Abyss-lord commented:
> That sounds like a better approach, and yes, checking metalake could give a better error message.

@justinmclean @xunliu 
 hi guys, Please help to see if the code structure is reasonable.
```java
public void handle() {
     if (!checkArguments){
         return;
     }
}

...

protected boolean checkArguments() {
    // check arguments and show 
    // combine displayMissingArguments and hasCompleteArguments
}
```
For the create command, the entity to be created will be validated. For example, the create metalake command will check the metalake parameter.

Additionally, it might be beneficial to further optimize the Command class by calling `execute` instead of `handle`. This would allow the argument checking logic to be moved up into the `Command` class itself, streamlining the process.
```java
public void execute(){
   // defined in Command class
   if (!checkArguments){ 
      return;
   }
   handle();
}

protected void handle() {
     // remain unchanged  
}

...

protected boolean checkArguments() {
    // check arguments and show 
    // combine displayMissingArguments and hasCompleteArguments
}
```

justinmclean commented:
This is going to result in a lot of code duplication. I would try to come up with another way.

Abyss-lord commented:
> This is going to result in a lot of code duplication. I would try to come up with another way. The solution suggested is also wider than the scope of the issue. The PR should focus on validating the entity's name.

@justinmclean  thank you Justin, in this PR, I will change the code at the handleXXXCommand layer to resolve the issue，If you come up with a better framework, you can assign the task to me.

justinmclean commented:
One possible way is to make subclass commands for each entity, as the validation for each entity type is the same. That I think, would reduce the duplication, but there might be something I'm missing. Let's try doing it at the handleXXXCommand layer and see what it looks like.

Abyss-lord commented:
> One possible way is to make subclass commands for each entity, as the validation for each entity type is the same. That I think, would reduce the duplication, but there might be something I'm missing. Let's try doing it at the handleXXXCommand layer and see what it looks like.

@justinmclean hi Justin, plz review this PR again when you have time. thank you.

xunliu reviewed (dismissed):
I think this PR focuses on solving  the Paramter validation of Table's list operation, which is still improved the CLI.
look good for my side.
@justinmclean What's do you think.

justinmclean commented:
There still a pending comment that needs to be address before this can be merged.

Abyss-lord commented:
> There are still two pending review comments that need to be addressed before this can be merged.

Hi, @justinmclean I just resolved all the comments, plz check again, thank you.

justinmclean commented:
@Abyss-lord I can still see two:
1. The variable missArguments could have a better name.
2. Please don't use MetadataObject.Type.xxx.name()

Abyss-lord commented:
> @Abyss-lord I can still see two:
> 
> 1. The variable missArguments could have a better name.
> 2. Please don't use MetadataObject.Type.xxx.name()

fix

justinmclean reviewed (approved):
LGTM thanks
